### PR TITLE
Bringing back the pub/sub binding for V2

### PR DESCRIPTION
**Pull request by nlopezgi**
Creating a V2 version of a pub/sub binding by refactoring the existing bits of V1 pub/sub and using as sample the existing binding for http/nats/sarama/kafka.

### CODE REVIEW

**Comment by nlopezgi** [comment]
1st draft towards fixing https://github.com/google/knative-gcp/issues/629

**Comment by n3wscott** [comment]
SWEEET note that you have to sign your commits with `git commit -s`

**Comment by n3wscott** [comment]
looks good, the build is red because it thinks you need to run the formatter. 

LGTM with the nit on formatter.

File: v2/protocol/pubsub/write_pubsub_message.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	b.Attributes = make(map[string]string)
+	return nil
+}
+
+func (b *pubsubMessagePublisher) End(ctx context.Context) error {
+	return nil
+}
+
+func (b *pubsubMessagePublisher) SetData(reader io.Reader) error {
+	val, err := ioutil.ReadAll(reader)
```
Comment (slinkydeveloper) [inline v2/protocol/pubsub/write_pubsub_message.go]: Avoid `ioutil.ReadAll` when possible, it's really costly. Use `bytes.Buffer` + `io.Copy` which optimizes use cases when the initial reader was already a buffer: https://github.com/cloudevents/sdk-go/blob/master/v2/protocol/kafka_sarama/write_producer_message.go#L59

File: v2/protocol/pubsub/message.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+	format   format.Format
+	version  spec.Version
+	encoding binding.Encoding
+}
+
+// NewMessage returns a binding.Message with data and attributes.
+// This message *can* be read several times safely
+func NewMessage(data []byte, attributes map[string]string) *Message {
+	encoding := binding.EncodingBinary
```
Comment (slinkydeveloper) [inline v2/protocol/pubsub/message.go]: What if the message is a non-cloudevent? In that case, this function should return `binding.EncodingUnknown`

File: v2/protocol/pubsub/message.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	// Structured v0.3
+	if v, ok := raw["specversion"]; ok {
+		var ver string
+		if err := json.Unmarshal(v, &ver); err != nil {
+			version = specs.Version("")
+		}
+		version = specs.Version(ver)
+	}
+	return &Message{
+		Msg:      pubsub.Message{Data: data, Attributes: attributes},
```
Comment (slinkydeveloper) [inline v2/protocol/pubsub/message.go]: Why you pack and unpack the data structure here? Can you just save `Data` and `Attributes` in the `Message` struct, without repacking them into `pubsub.Message`?

File: v2/protocol/pubsub/message.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	var version spec.Version = nil
+	if attributes != nil {
+		// Use Content-type attr to determine if message is structured and
+		// set format.
+		if s := attributes[contentType]; format.IsFormat(s) {
+			f = format.Lookup(s)
+			encoding = binding.EncodingStructured
+		}
+		// Binary v0.3:
+		if s := attributes[prefix+"specversion"]; s != "" {
```
Comment (slinkydeveloper) [inline v2/protocol/pubsub/message.go]: `prefix+"specversion"` -> `specs.PrefixedSpecVersionName()`

File: v2/protocol/pubsub/message.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+		version = specs.Version("")
+	}
+
+	// Structured v0.3
+	if v, ok := raw["specversion"]; ok {
+		var ver string
+		if err := json.Unmarshal(v, &ver); err != nil {
+			version = specs.Version("")
+		}
+		version = specs.Version(ver)
```
Comment (slinkydeveloper) [inline v2/protocol/pubsub/message.go]: Why do you unmarshal the json to read the version? you don't need that

File: v2/protocol/pubsub/message.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		}
+		// Binary v0.3:
+		if s := attributes[prefix+"specversion"]; s != "" {
+			version = specs.Version(s)
+		}
+	}
+
+	// Check as Structured encoding.
+	raw := make(map[string]json.RawMessage)
+	if err := json.Unmarshal(data, &raw); err != nil {
```
Comment (slinkydeveloper) [inline v2/protocol/pubsub/message.go]: Why you do that? This check is done directly when you read or buffer the message

File: v2/protocol/pubsub/message.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+var specs = spec.WithPrefix(prefix)
+
+// Message represents a Pub/Sub message.
+// This message *can* be read several times safely
+type Message struct {
+	Msg pubsub.Message
+
+	format   format.Format
+	version  spec.Version
+	encoding binding.Encoding
```
Comment (slinkydeveloper) [inline v2/protocol/pubsub/message.go]: You can have `format` and `version` mutually exclusive to identify the encoding, without saving it as field of the struct: https://github.com/cloudevents/sdk-go/blob/master/v2/protocol/kafka_sarama/message.go#L78

File: v2/protocol/pubsub/protocol.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+		if err := fn(t); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Send implements Sender.Send
+func (t *Protocol) Send(ctx context.Context, in binding.Message) error {
+	topic := cecontext.TopicFrom(ctx)
```
Comment (slinkydeveloper) [inline v2/protocol/pubsub/protocol.go]: `Sender` & `Requester` contract requires that `Send` invokes `Finish` on the input message

File: v2/protocol/pubsub/message.go (hunk 7 above)
Comment (nlopezgi) [reply v2/protocol/pubsub/message.go]: done

File: v2/protocol/pubsub/message.go (hunk 3 above)
Comment (nlopezgi) [reply v2/protocol/pubsub/message.go]: done

File: v2/protocol/pubsub/protocol.go (hunk 8 above)
Comment (nlopezgi) [reply v2/protocol/pubsub/protocol.go]: done

File: v2/protocol/pubsub/write_pubsub_message.go (hunk 1 above)
Comment (nlopezgi) [reply v2/protocol/pubsub/write_pubsub_message.go]: done

File: v2/protocol/pubsub/message.go (hunk 4 above)
Comment (nlopezgi) [reply v2/protocol/pubsub/message.go]: done

File: v2/protocol/pubsub/message.go (hunk 5 above)
Comment (nlopezgi) [reply v2/protocol/pubsub/message.go]: this was code that I ported from the previous implementation of message.go. Removed it.

File: v2/protocol/pubsub/message.go (hunk 6 above)
Comment (nlopezgi) [reply v2/protocol/pubsub/message.go]: this was code that I ported from the previous implementation of message.go. Removed it.

File: v2/protocol/pubsub/message.go (hunk 2 above)
Comment (nlopezgi) [reply v2/protocol/pubsub/message.go]: removed encoding as field and used same strategy as https://github.com/cloudevents/sdk-go/blob/master/v2/protocol/kafka_sarama/message.go#L78, iiuc, this should now be done properly.

**Comment by nlopezgi** [comment]
@nachocano let me know if you want me to wait for your review of this PR or if its good to go once @slinkydeveloper and @n3wscott sign off.

**Comment by nachocano** [comment]
Sorry, missed this. But it's fine to go in!
Thanks Nick and guys for reviewing!

**Pull request merged**